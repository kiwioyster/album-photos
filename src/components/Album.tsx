import React, { useState } from 'react';
import axios from 'axios';
import AlbumPhoto from './AlbumPhoto';
import { IAlbum, IUser, IPhoto } from '../type';

interface Props {
  album: IAlbum;
  users: IUser[];
}

const Album: React.FC<Props> = ({ album, users }) => {
  const [photos, setPhotos] = useState<IPhoto[]>([]);
  const PHOTO_URL = 'http://jsonplaceholder.typicode.com/photos';
  const fetchPhoto = async (albumId: number) => {
    try {
      const response = await axios.get<IPhoto[]>(
        `${PHOTO_URL}?albumId=${albumId}`
      );
      setPhotos(response.data);
    } catch (e) {
      console.error(e);
    }
  };
  const clickedAlbum = (id: number) => {
    fetchPhoto(id);
  };
  return (
    <div className='album' onClick={() => clickedAlbum(album.id)}>
      <div className='album-title'>Title: {album.title}</div>
      <div className='album-user'>
        User: {users.find((user) => user.id === album.userId)?.name}
      </div>
      <AlbumPhoto photos={photos} />
    </div>
  );
};

export default Album;
