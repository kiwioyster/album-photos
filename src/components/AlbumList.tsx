import React, { useState, useEffect } from 'react';
import axios from 'axios';
import Album from './Album';
import { IAlbum, IUser } from '../type';

const AlbumList: React.FC<{}> = () => {
  const [albums, setAlbums] = useState<IAlbum[]>([]);
  const [users, setUsers] = useState<IUser[]>([]);
  const ALBUMS_URL = 'http://jsonplaceholder.typicode.com/albums';
  const USERS_URL = 'http://jsonplaceholder.typicode.com/users';
  const fetchAlbums = async () => {
    try {
      const response = await axios.get(ALBUMS_URL);
      setAlbums(response.data);
    } catch (e) {
      console.error(e);
    }
  };
  const fetchUsers = async () => {
    try {
      const response = await axios.get(USERS_URL);
      setUsers(response.data);
    } catch (e) {
      console.error(e);
    }
  };
  useEffect(() => {
    fetchAlbums();
    fetchUsers();
  }, []);

  return (
    <ul>
      {albums.map((album, i) => {
        return (
          <li key={i}>
            <Album album={album} users={users} />
          </li>
        );
      })}
    </ul>
  );
};

export default AlbumList;
