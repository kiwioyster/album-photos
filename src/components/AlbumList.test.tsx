import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import AlbumList from './AlbumList';
import Adapter from 'enzyme-adapter-react-16';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { act, cleanup } from '@testing-library/react';
import { IAlbum, IUser } from '../type';
import Album from './Album';

configure({ adapter: new Adapter() });
describe('AlbumList', () => {
  afterEach(cleanup);
  it('should render', () => {
    const component = shallow(<AlbumList />);
    expect(component.getElements()).toMatchSnapshot();
  });
  it('should get from api', async () => {
    const mock = new MockAdapter(axios);
    const mockAlbums: IAlbum[] = [
      {
        userId: 1,
        id: 1,
        title: 'mock title ',
      },
      {
        userId: 1,
        id: 2,
        title: 'mock title 2',
      },
      {
        userId: 2,
        id: 3,
        title: 'mock title 3',
      },
    ];
    const mockUsers: IUser[] = [
      {
        id: 1,
        name: 'mock user',
        username: 'mock username',
        address: 'mock address',
        phone: 'mock phone',
        website: 'mock website',
        company: 'mock company',
      },
      {
        id: 2,
        name: 'mock user 2',
        username: 'mock username 2',
        address: 'mock address 2',
        phone: 'mock phone 2',
        website: 'mock website 2',
        company: 'mock company 2',
      },
    ];
    mock
      .onGet('http://jsonplaceholder.typicode.com/albums')
      .reply(200, mockAlbums);
    mock
      .onGet('http://jsonplaceholder.typicode.com/users')
      .reply(200, mockUsers);
    const component = mount(<AlbumList />);
    await act(
      () =>
        new Promise<void>((resolve) => {
          setImmediate(() => {
            component.update();
            resolve();
          });
        })
    );
    expect(component.find(Album).length).toEqual(3);
    expect(component.find(Album).at(0).prop('album')).toEqual(mockAlbums[0]);
    expect(component.find(Album).at(0).prop('users')).toEqual(mockUsers);
    expect(component.find(Album).at(1).prop('album')).toEqual(mockAlbums[1]);
    expect(component.find(Album).at(1).prop('users')).toEqual(mockUsers);
    expect(component.find(Album).at(2).prop('album')).toEqual(mockAlbums[2]);
    expect(component.find(Album).at(2).prop('users')).toEqual(mockUsers);
  });
});
