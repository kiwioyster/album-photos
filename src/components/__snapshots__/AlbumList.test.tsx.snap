// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumList should render 1`] = `
Array [
  <ul />,
]
`;
